### PR TITLE
Add logic to remove default class_transformer.excludes when security …

### PR DESCRIPTION
**Pull request by jasonjkeller**
…agent is enabled

DO NOT MERGE!

---

By default, the following three `class_transformer.exclude` rules defined in `META-INF/excludes` will be applied to prevent any classes from being transformed in packages matching `^java/security/.*`, `^javax/crypto/.*`, or `^net/sf/saxon.*`. Exclude rules are evaluated when determining whether to transform classes in the `InstrumentationContextManager` or the `PointCutClassTransformer`.

```
### All default excludes listed below get ignored when the security agent is enabled
# exclude java.security and sun.reflect classes from transformation
^java/security/.*
# crypto classes can cause class circularity errors if they get too far along in the class transformer
^javax/crypto/.*
# saxon xlst classes - see ticket 195949
^net/sf/saxon.*
```

Classes from these packages are excluded from transformation to prevent various issues such as long startup times (`net/sf/saxon`), `ClassCircularityError`s (`javax/crypto`), and `StackOverflowError`s (`java/security`).

However, preventing transformation of classes in these packages hinders functionality of the security agent.

This PR will remove the three listed default exclude rules when the security agent is enabled. The following will be logged indicating that this behavior is taking place.

```
2023-08-16T15:17:11,001-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: Ignored ^java/security/.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
2023-08-16T15:17:11,001-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: Ignored ^javax/crypto/.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
2023-08-16T15:17:11,001-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: Ignored ^net/sf/saxon.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
2023-08-16T15:17:11,012-0700 [43845 1] com.newrelic FINER: Ignored ^java/security/.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
2023-08-16T15:17:11,012-0700 [43845 1] com.newrelic FINER: Ignored ^javax/crypto/.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
2023-08-16T15:17:11,012-0700 [43845 1] com.newrelic FINER: Ignored ^net/sf/saxon.* class_transformer exclude defined in META-INF/excludes because the security agent is enabled. This can be overridden by explicitly setting newrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.* or NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
```

Additionally, even if the security agent is enabled, these exclude rules can be explicitly added via one of the following agent config options, in which case the user configured exclude rule will take precedence and will not be removed due to the security agent being enabled.

YAML:
```
  class_transformer:
    excludes: ^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
```

System property:
```
-Dnewrelic.config.class_transformer.excludes=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
```

Environment variable:
```
NEW_RELIC_CLASS_TRANSFORMER_EXCLUDES=^javax/crypto/.*,^java/security/.*,^net/sf/saxon.*
```

When explicitly setting the excludes via user config the following will be logged:

```
2023-08-16T15:17:11,000-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: ^java/security/.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
2023-08-16T15:17:11,000-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: ^javax/crypto/.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
2023-08-16T15:17:11,000-0700 [43845 1] com.newrelic.agent.instrumentation.PointCutClassTransformer FINER: ^net/sf/saxon.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
2023-08-16T15:17:11,011-0700 [43845 1] com.newrelic FINER: ^java/security/.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
2023-08-16T15:17:11,011-0700 [43845 1] com.newrelic FINER: ^javax/crypto/.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
2023-08-16T15:17:11,012-0700 [43845 1] com.newrelic FINER: ^net/sf/saxon.* class_transformer exclude explicitly added by user config. The user configured exclude rule will take precedence and will not be ignored due to the security agent being enabled.
```

### CODE REVIEW

File: newrelic-agent/src/main/java/com/newrelic/agent/config/SecurityAgentConfig.java [hunk 1]
```diff
@@ -44,6 +48,8 @@ public class SecurityAgentConfig {
     private static final Config config = NewRelic.getAgent().getConfig();
     private static final String ENABLED = "enabled";
     private static final String DISABLED = "disabled";
+    public static final Set<String> SECURITY_AGENT_CLASS_TRANSFORMER_EXCLUDES_TO_IGNORE = Collections.unmodifiableSet(
+            Sets.newHashSet("^java/security/.*", "^javax/crypto/.*", "^net/sf/saxon.*"));
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/config/SecurityAgentConfig.java:52]: This is the full set of default `class_transformer.exclude` rules that need to be ignored when the security agent is enabled. Should further exclude rules need to be treated the same in the future they can simply be added here and no other logic should need to be changed elsewhere in the agent codebase. Only new tests would need to be added to `IncludeExcludeTest`.

File: newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/context/InstrumentationContextManager.java [hunk 2]
```diff
@@ -221,11 +220,6 @@ public boolean shouldTransform(final String internalClassName, final ClassLoader
             Agent.LOG.log(Level.FINEST, "Skipping transform of {0}. Classloader {1} is excluded.", internalClassName, classloader);
             return false;
         }
-        if (internalClassName.startsWith("javax/crypto/")) {
-            // crypto classes can cause class circularity errors if they get too far along in the class transformer
-            Agent.LOG.finest(MessageFormat.format("Instrumentation skipped by ''javax crypto'' rule: {0}", internalClassName));
-            return false;
-        }
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/context/InstrumentationContextManager.java:228]: This exclude rule was moved to the `META-INF/excludes` file to simplify the logic and keep all of the security agent related excludes in a central place.

File: newrelic-agent/src/main/resources/META-INF/excludes [hunk 3]
```diff
@@ -88,4 +84,11 @@
...
-^com/ibm/ws/webcontainer/servlet/ServletWrapperImpl
\ No newline at end of file
+^com/ibm/ws/webcontainer/servlet/ServletWrapperImpl
+### All default excludes listed below get ignored when the security agent is enabled
+# exclude java.security and sun.reflect classes from transformation
+^java/security/.*
+# crypto classes can cause class circularity errors if they get too far along in the class transformer
+^javax/crypto/.*
+# saxon xlst classes - see ticket 195949
+^net/sf/saxon.*
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/resources/META-INF/excludes:94]: These are the three default exclude rules that are effected by the security agent.

File: newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/ClassNameFilter.java [hunk 4]
```diff
@@ -112,15 +124,34 @@ public void addExcludeFileClassFilters() {
                 // ignore
             }
         }
+        boolean ignoreExcludeForSecurityAgent = false;
+        String formattedIgnoredExcludes = String.join(",", SECURITY_AGENT_CLASS_TRANSFORMER_EXCLUDES_TO_IGNORE);
+        boolean shouldInitializeSecurityAgent = shouldInitializeSecurityAgent();
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/ClassNameFilter.java:129]: Only consider removing one of the default exclude rules if the `shouldInitializeSecurityAgent()` check evaluates to `true`.

**Review by jtduffy — APPROVED** [review]

**Comment by lovesh-ap** [comment]
@jasonjkeller Hey there! 👋 Just wanted to let you know that we currently have an open issue [(#154979)](https://issues.newrelic.com/browse/NR-154979) related to the changes in this PR. It might be good to keep an eye on that while reviewing this. Thanks!

**Review by tbradellis — APPROVED** [review]

**Pull request merged**